### PR TITLE
Unit testing for workspace uploader

### PR DESCRIPTION
**Pull request by Xiangs18**
Screen Shot 2023-08-17 at 3 49 38 PM

### CODE REVIEW

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 1]
```diff
@@ -1,5 +1,209 @@
...
+
+    yield tmp_dir, assembly_dirs, assembly_names, upas, workspace_id, upload_env_key, target_files
+
+    shutil.rmtree(tmp_dir, ignore_errors=True)
+
+
+def test_get_yaml_file_path(setup_and_teardown):
+    _, assembly_dirs, _, _, _, _, _ = setup_and_teardown
+    yaml_path = workspace_uploader._get_yaml_file_path(assembly_dirs[0])
+
```
Comment (Tianhao-Gu) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:52]:
```suggestion
assert yaml_path.endswith(UPLOADED_YAML)
```

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 2]
```diff
@@ -1,5 +1,209 @@
...
+
+def test_read_upload_status_yaml_file(setup_and_teardown):
+    _, assembly_dirs, assembly_names, _, workspace_id, upload_env_key, _ = setup_and_teardown
+    assembly_dir = assembly_dirs[0]
+    assembly_name = assembly_names[0]
+
+    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
+        upload_env_key, workspace_id, assembly_dir, assembly_name
+    )
```
Comment (Tianhao-Gu) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:73]: can you comment `test empty yaml file in assembly_dir`?

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 1 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:52]: I already asserted `os.path.exists(yaml_path)` in line 53. Do I need to add this additional check?

Comment (Tianhao-Gu) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:52]: 👍

**Review by Tianhao-Gu — APPROVED** [review]
LGTM

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 3]
```diff
@@ -1,5 +1,210 @@
...
+        for assembly_name, assembly_dir in zip(assembly_names, assembly_dirs)
+    }
+
+    conf = Mock()
+    username = "kbase"
+    job_dir = loader_helper.make_job_dir(tmp_dir, username)
+    conf.job_data_dir = loader_helper.make_job_data_dir(job_dir)
+    data_dir = workspace_uploader._prepare_skd_job_dir_to_upload(conf, wait_to_upload_assemblies)
+    
+    assert sorted(os.listdir(data_dir)) == sorted(assembly_names)
```
Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:135]: Should check the hardlinks were made as expected

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 4]
```diff
@@ -1,5 +1,210 @@
...
+    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
+        upload_env_key, workspace_id, host_assembly_dir, assembly_name
+    )
+    expected_data = {
+        upload_env_key: {workspace_id: {"file_name": assembly_name, "upa": upa}}
+    }
+
+    assert uploaded
+    assert expected_data == data
+    assert os.readlink(os.path.join(upload_dir, upa)) == os.path.join(output_dir, upa)
```
Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:168]: Should check the file hardlink as well

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 5]
```diff
@@ -1,5 +1,210 @@
...
+
+
+def test_upload_assembly_to_workspace(setup_and_teardown):
+    _, _, assembly_names, upas, workspace_id, _, _  = setup_and_teardown
+    upa = upas[0]
+    assembly_name = assembly_names[0]
+    file_path = "/path/to/file/in/AssembilyUtil"
+
+    conf = Mock()
+    conf.asu.save_assembly_from_fasta2.return_value = {"upa": upa}
```
Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:178]: I would use `create_autospec` here for `asu` rather than a bare mock
https://m1lt0n.github.io/python/unittest/mock/autospec/mocking-with-autospec/
https://docs.python.org/3/library/unittest.mock.html#auto-speccing

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 6]
```diff
@@ -1,5 +1,210 @@
...
+    upa_res = workspace_uploader._upload_assembly_to_workspace(
+        conf, workspace_id, file_path, assembly_name
+    )
+
+    expected_upa = upa.replace("/", "_")
+    assert expected_upa == upa_res
+    assert conf.asu.save_assembly_from_fasta2.call_count == 1
```
Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:185]: Why not check the arguments are as expected?

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 7]
```diff
@@ -1,5 +1,210 @@
...
+    conf.output_queue = Queue()
+    for assembly_name, upa in zip(assembly_names, upas):
+        conf.output_queue.put((assembly_name, upa))
+    
+    failed_names = workspace_uploader._upload_assembly_files_in_parallel(
+        conf, upload_env_key, workspace_id, upload_dir, wait_to_upload_assemblies
+    )
+
+    assert conf.output_queue.empty()
+    assert failed_names == []
```
Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:210]: Why not check the input queue contents are as expected?

Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:210]: A similar test where an upload failed would be useful

**Review by MrCreosote — COMMENTED** [review]
So usually unit tests shouldn't be accessing internal APIs [1], but given that real testing would require contacting kbase services, using real assembly files, etc. this seems reasonable for now, although not comprehensive.

When we get to the serious test writing stage we'll probably want to stand up the data services in a text fixture and run full tests, although to my mind we might want to do the data base conversion first.

[1] https://blog.ploeh.dk/2015/09/22/unit-testing-internals/

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 8]
```diff
@@ -1,5 +1,210 @@
...
+                      "GCF_000979175.1_gtlEnvA5udCFS_genomic.fna.gz"]
+
+    assembly_dirs, target_files = list(), list()
+    for assembly_dir_name, assembly_name in zip(assembly_dir_names, assembly_names):
+        target_dir_path = sourcedata_dir.joinpath(assembly_dir_name)
+        target_dir_path.mkdir()
+        target_file_path = target_dir_path.joinpath(assembly_name)
+        target_file_path.touch()
+        new_dir_path = collection_source_dir.joinpath(assembly_dir_name)
+        os.symlink(target_dir_path.resolve(), new_dir_path.resolve(), target_is_directory=True)
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:40]: because upa_dirs in collectionssource softlink to upa_dirs in sourcedata. This is the current setup.

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 9]
```diff
@@ -1,5 +1,210 @@
...
+    collection_source_dir = Path(tmp_dir) / "collectionssource"
+    sourcedata_dir.mkdir(parents=True)
+    collection_source_dir.mkdir()
+
+    workspace_id = 12345
+    upload_env_key = "CI"
+    upas = ["12345_58_1", "12345_58_2"]
+    assembly_dir_names = ["GCF_000979855.1", "GCF_000979175.1"]
+    assembly_names = ["GCF_000979855.1_gtlEnvA5udCFS_genomic.fna.gz",
+                      "GCF_000979175.1_gtlEnvA5udCFS_genomic.fna.gz"]
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:31]: > or better yet, create them in the tests and have them differ from test to test where possible (e.g. not for assembly_*_names)

Why so?

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:40]: Oh, that's the symlink created by the GTDB/NCBI downloader?

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:31]: There's less of a chance to catch bugs if the test all use the same inputs. Bigger variety of inputs = bigger chance to catch edge cases

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:40]: yeah, https://github.com/kbase/collections/blob/deeecb44b01497c51c4efb68a903cbe623457f46/src/loaders/ncbi_downloader/gtdb.py#L246

[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:40]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:31]: Make sense. 

But assembly_names are connected to assembly_dir_names from the GTDB/NCBI downloader. So i guess is okay to pass assembly_names to every test as well?

[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:31]: If it were me I'd just make them constants but I don't feel particularly strongly about it. Up to you

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 10]
```diff
@@ -1,5 +1,210 @@
...
+
+    expected_data = {
+        upload_env_key: {workspace_id: {"file_name": assembly_name, "upa": upa}}
+    }
+
+    assert uploaded
+    assert expected_data == data
+
+
+def test_fetch_assemblies_to_upload(setup_and_teardown):
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 4 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:212]: Added hard link check

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 6 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:231]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 11]
```diff
@@ -1,5 +1,210 @@
...
+    yield tmp_dir, assembly_dirs, assembly_names, upas, workspace_id, upload_env_key, target_files
+
+    shutil.rmtree(tmp_dir, ignore_errors=True)
+
+
+def test_get_yaml_file_path(setup_and_teardown):
+    _, assembly_dirs, _, _, _, _, _ = setup_and_teardown
+    yaml_path = workspace_uploader._get_yaml_file_path(assembly_dirs[0])
+
+    assert os.path.exists(yaml_path)
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Made assembly_dir_names and assembly_names constants, and have others differ from test to test

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 3 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 7 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Updated!

Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: added an upload failed

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 5 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 12]
```diff
@@ -1,5 +1,210 @@
...
+        target_dir_path = sourcedata_dir.joinpath(assembly_dir_name)
+        target_dir_path.mkdir()
+        target_file_path = target_dir_path.joinpath(assembly_name)
+        target_file_path.touch()
+        new_dir_path = collection_source_dir.joinpath(assembly_dir_name)
+        os.symlink(target_dir_path.resolve(), new_dir_path.resolve(), target_is_directory=True)
+        assembly_dirs.append(str(new_dir_path))
+        target_files.append(str(target_file_path))
+
+    yield tmp_dir, assembly_dirs, assembly_names, upas, workspace_id, upload_env_key, target_files
```
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Looks good but I'd move it out of the function to a module property

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: i wouldn't get rid of the existence check, both are useful

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 3 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 4 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:212]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 5 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: You should  add `spec_set=True, instance=True` like this:
https://github.com/kbase/sample_service/blob/9ffd516a66d5dc046aa48a339560b158928eb2c0/test/core/workspace_test.py#L288

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 6 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:231]: Sorry, I'm not seeing the change for this one? It looks like it's still just testing the call count, not the arguments. You'd need to do something like
```python
conf.asu.save_assembly_from_fasta2.assert_called_once_with(<args go here>)
```

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 7 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: I'm not seeing the already uploaded assembly test?

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: I don't know. I had it here though: https://github.com/kbase/collections/blob/b38b38ba39bf38800e0fef92acd91a138e4608e2/test/src/loaders/workspace_uploader/workspace_uploader_test.py#L125-L128

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 13]
```diff
@@ -1,5 +1,302 @@
...
+    # an uploaded successful
+    conf.output_queue.put((ASSEMBLY_NAMES[0], upa))
+    # an upload failed
+    conf.output_queue.put((ASSEMBLY_NAMES[1], None))
+
+    failed_names = workspace_uploader._upload_assembly_files_in_parallel(
+        conf, upload_env_key, workspace_id, upload_dir, wait_to_upload_assemblies
+    )
+
+    (
```
Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:263]: What I would do here, rather than testing each item of the queue tuple once by one, is make expected tuples and then 

```
assert conf.input_queue.get() == expected_tuple1
assert conf.input_queue.get() == expected_tuple2
```

pytest is good about providing error messages showing where the differences in data structures exist

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py [hunk 14]
```diff
@@ -1,5 +1,302 @@
...
+        upload_env_key: {workspace_id: {"file_name": assembly_name, "upa": upa}}
+    }
+
+    assert uploaded
+    assert expected_data == data
+
+    with pytest.raises(ValueError, match=f"already exists in workspace"):
+        workspace_uploader._update_upload_status_yaml_file(
+            upload_env_key, workspace_id, upa, assembly_dir, assembly_name
+        )
```
Comment (MrCreosote) [inline test/src/loaders/workspace_uploader/workspace_uploader_test.py:128]: You can inline a lot of this stuff - this is just an example

```suggestion
    assembly_dir = params.assembly_dirs[0]
    assembly_name = ASSEMBLY_NAMES[0]

    workspace_uploader._update_upload_status_yaml_file(
        "PROD", 42, "34_2_7", assembly_dir, assembly_name
    )
    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
        "PROD", 42, assembly_dir, assembly_name
    )

    expected_data = {
        "PROD": {42: {"file_name": assembly_name, "upa": "34_2_7"}}
    }

    assert uploaded
    assert expected_data == data

    with pytest.raises(ValueError, match=f"already exists in workspace"):
        workspace_uploader._update_upload_status_yaml_file(
            "PROD", 42, "34_2_7", assembly_dir, assembly_name
        )
```

**Review by MrCreosote — DISMISSED** [review]

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: That's on the previous test

[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: Perfectly fine to have that though

[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: ah, I am not sure. This is like what we currently have on dev-unit_testing branch. Is right there and I have not push any new commits.

https://github.com/kbase/collections/blob/dev-unit_testing/test/src/loaders/workspace_uploader/workspace_uploader_test.py

[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: I see the same thing there. `test_update_upload_status_yaml_file` has a failing section but `test_fetch_assemblies_to_upload` doesn't, and there's no new test to exercise the failing case

[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: I see. You want me to add a failing case on `test_fetch_assemblies_to_upload `.

[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:131]: Right, or just add a whole new test for the failing case, whichever makes more sense

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 5 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Is there a workaround can help me get rid of AttributeError: Mock object has no attribute 'asu' error?

Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Use a regular `Mock` for `Conf`. `create_autospec` doesn't know about dynamically created properties like `asu`

Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: So something like
```
conf = Mock()
conf.asu = create_autospec(...)
```

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 14 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:128]: What are the benefits doing this? I kinda feel inline is more error-prone.

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]:
```
class Params(NamedTuple):
    tmp_dir: str
    sourcedata_dir: str
    collection_source_dir: str
    assembly_dirs: list[str]
    target_files: list[str]
```

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:70]: Updated

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:129]: Updated

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 5 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: works like a charm. Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 6 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 13 above)
Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 14 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:126]: * The big on is that you're now testing based on equality rather than identity. This is more important for classes but it's a good habit to get into.
* fewer test lines
* easier to read the tests naturally
* easier to check that your tests work by forcing them to fail

Comment (Xiangs18) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: Updated!

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:70]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py:121]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 5 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 6 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 13 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

File: test/src/loaders/workspace_uploader/workspace_uploader_test.py (hunk 14 above)
Comment (MrCreosote) [reply test/src/loaders/workspace_uploader/workspace_uploader_test.py]: :+1:

**Review by MrCreosote — COMMENTED** [review]
LGTM

**Review by MrCreosote — APPROVED** [review]
*&)&_&))^^Y&*%(
